Add new openpolicyagent connector (V2)

rohithdv opened the pull request:
This PR fixes #702.
This PR has the following changes: 
(1) Adds a `lib` folder in `connectors/open_policy_agent` which has functionality to contact catalog connector and opa to retrieve policy decisions. This folder is based on the `connectors/opa/lib` folder currently used in the GRPC based opa connector. 

(2) Updates the `connectors/open_policy_agent/testutil `to be compatible with openapi taxonomy related changes. 

(3) changes the codebase in `connectors/open_policy_agent/lib` to be compatible with openapi taxonomy related changes. 

(4) Updates the `api_default.go` in `connectors/open_policy_agent/go` to have functionality of contacting OPA once the Openapi based HTTP Request arrives in the gin based http server (which is the V2 opa connector) 

(5) `GetPoliciesDecisions` function in `api_default.go` now returns back an HTTP response which contains the policy manager response object (which is compatible with the openapi based taxonomy). 

(6) Added `charts/fybrik/templates/openpolicyagent-connector` folder which has the helm based files to start V2 OPA connector (Openapi based HTTP server ) during the install of fybrik (i.e. when the command  `helm install fybrik charts/fybrik --set global.tag=latest -n fybrik-system --wait`) is run. 

**PS:** Currently during fybrik installation, we are deploying V1 and V2 versions of OPA Connector to check compatibility with GRPC and OpenAPI protocols. We will remove the V1 opa connector in a subsequent PR once the end to end flow of Openapi based taxonomy is fully tested. 

To summarize at a high level, currently the fybrik deployment looks like below:

`pod/openpolicyagent-connector-84bfc6f495-2z5n6 `refers to V2 OPA connector (Openapi HTTP based)
`pod/opa-connector-674d8d489d-rmwqg`  refers to V1 OPA Connector (GRPC based)
`pod/manager-client-sn7bz`  is a test client which mocks the manager functionality related to openapi taxonomy changes. The `manager-client`  is a temporary installation only for testing purposes and will be removed in a cleanup PR. 

![image](https://user-images.githubusercontent.com/59818576/130641964-bd538915-4c9e-4866-8212-b5bd788a8ab4.png)

roee88 commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,109 @@
+	return &input, nil
+}
+
+func displayReq(input2 openapiclientmodels.PolicyManagerRequest) error {
+	bytes, err := input2.MarshalJSON()
+	if err != nil {
+		log.Println("err: ", err)
+		return nil
+	}
+	log.Println("marshalled request:", string(bytes))
+	return nil
+}
```
I believe we implemented such debug functions several times already. I'm not even sure if it's needed and printing with %+v would be enough but in any case a function like that should move to `pkg/` and be reused across the project. Look if we have an implementation already in manage (cc @shlomitk1). 

I'm personally not a fan of these functions since they require marshalling just for debugging. I would prefer to use a gin middleware instead for logging both request and response and add a flag to either add the middleware or not.

rohithdv replied:
@roee88 is it ok to use this library [render](https://github.com/luci/go-render) as suggested [here](https://stackoverflow.com/questions/43259971/golang-how-to-print-struct-value-with-pointer) ? Currently I have used this library in the opa connector as shown [here](https://github.com/rohithdv/fybrik/blob/186fcd6c2bc818754c825f777e9fe0843c5eaf27/connectors/open_policy_agent/go/api_default.go#L57).

or do you suggest some other way for printing structs with pointers ? Please advise. Thanks.

ritwikc commented:
Just a few questions/comments I am putting here.:
1. In opa_reader.go do you want to print the tags and the values ? I know those are for debugging. But are those needed in the merged code.
2. In  test/services/pilot/mock-openapi-policymanager/manager-client.go I see the <SECRET-NAME> and <NAMESPACE> are hard coded.
3. In charts/fybrik/values.yaml on line 475 you are referring to rohithdv. Is that what it is supposed to be ?

rohithdv commented on charts/fybrik/templates/openpolicyagent-connector/openpolicyagent-connector-cm.yaml:
```diff
@@ -0,0 +1,13 @@
+{{- $autoFlag := and .Values.coordinator.enabled (eq .Values.coordinator.policyManager "opa") }}
```
Fixed. Please check.

rohithdv commented on charts/fybrik/templates/openpolicyagent-connector/openpolicyagent-connector-deployment.yaml:
```diff
@@ -0,0 +1,63 @@
+          imagePullPolicy: {{ .Values.openPolicyAgentConnector.imagePullPolicy | default .Values.global.imagePullPolicy }}
+          ports:
+            - name: http
+              containerPort: 8081
```
Fixed. Please check.

rohithdv commented on connectors/open_policy_agent/Dockerfile:
```diff
@@ -0,0 +1,7 @@
+
+COPY openpolicyagent-connector /
+
+EXPOSE 50082
```
Fixed. Please check.

rohithdv commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,109 @@
+	}
+	log.Printf("Env. variable extracted: %s - %s\n", key, value)
+	return value
+}
```
I have changed the implementation. Please review. Thanks.

rohithdv commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,109 @@
+	log.Printf("input: %v", input)
+
+	return &input, nil
+}
```
Fixed. Please check.

rohithdv commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,109 @@
+	log.Println("contactOPA: Received evaluation after execution of GetOPADecisions : " + string(jsonOutput))
+
+	return eval, nil
+}
```
Most of the statements in this function (not related to catalog connector and opa server) are now removed. I have kept the function still with the necessary modification. Please suggest if I need to alter anything else. Thanks.

rohithdv commented:
> 2\. see the  and  are hard code

Addressed all the comments except (1). I think the log statements are useful till we test end to end flow. Once that is done, we can prune unnecessary logs. Hope this is fine for now. Thanks.

roee88 commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,54 @@
-// GetPoliciesDecisions - getPoliciesDecisions
-func GetPoliciesDecisions(c *gin.Context) {
+type ConnectorController struct {
+	Policytobeevaluated string
+	Opaserverurl        string
+	Opaclient           *http.Client
```
Use CamelCase (current names are unreadable)

rohithdv replied:
fixed. Please check. Thanks.

roee88 commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,54 @@
-// GetPoliciesDecisions - getPoliciesDecisions
-func GetPoliciesDecisions(c *gin.Context) {
+type ConnectorController struct {
+	Policytobeevaluated string
```
Why do we need this?

rohithdv replied:
made Policytobeevaluated to be a constant now. Thanks.

roee88 commented on connectors/open_policy_agent/go/routers.go:
```diff
@@ -18,59 +18,15 @@ import (
 	return router
 }
 
 // Index is the index handler.
 func Index(c *gin.Context) {
 	c.String(http.StatusOK, "Hello World!")
 }
```
Remove?

rohithdv replied:
fixed. Please check. Thanks.

roee88 commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,54 @@
+}
+
+func (e *ConnectorController) contactOPA(input openapiclientmodels.PolicyManagerRequest, creds string) (openapiclientmodels.PolicyManagerResponse, error) {
+	opaReader := opabl.NewOpaReader(e.Opaserverurl)
```
Reuse opa client instead of creating a new one on each call

rohithdv replied:
fixed. Please check. Thanks.

roee88 commented on connectors/open_policy_agent/go/api_default.go:
```diff
@@ -13,18 +13,54 @@
+
+	eval, err := opaReader.GetOPADecisions(&input, creds, e.CatalogReader, e.Policytobeevaluated, e.Opaclient)
+	if err != nil {
+		log.Println("GetOPADecisions err:", err)
```
Don't log returned errors. You may use errors.Wrap instead if you need to add information to the error.

rohithdv replied:
fixed. Please check. Thanks.

rohithdv commented on charts/fybrik/values.yaml:
```diff
@@ -59,11 +59,11 @@ coordinator:
   # Overrides the policy manager connector URL.
   # Defaults to `<policyManager>-connector:80`.
-  policyManagerConnectorURL: ""
+  policyManagerConnectorURL: "http://openpolicyagent-connector:80"
```
fixed. Please check. Thanks.
